stories/2024-01-01-Eurovision/story.ipynb

```diff
@@ -2840,6 +2840,674 @@
+                "    test = test_set.copy()\n",
+                "\n",
+                "    test['prediction_rel'] = np.nan\n",
+                "    test['predictions'] = np.nan\n",
+                "\n",
+                "    for i in test['id'].unique():\n",
+                "        sub_test = test.loc[test['id'] == i, ~test.columns.isin(['id', 'rank', 'prediction_rel', 'predictions'])]\n",
+                "\n",
+                "        preds = model.predict(sub_test)\n",
+                "\n",
+                "        # get the order of the predictions\n",
+                "        res = rankdata(preds, method='ordinal')\n",
+                "        test.loc[test['id'] == i, 'prediction_rel'] = res\n",
+                "\n",
+                "        # create a score for top 10 predictions getting 1:10 and others 0\n",
+                "        top10 = test.loc[test['id'] == i, 'prediction_rel'].nlargest(10).values\n",
+                "\n",
+                "\n",
+                "        test.loc[test['id'] == i, 'predictions'] = test.loc[test['id'] == i, 'prediction_rel'].apply(lambda x: x if x in top10 else 0)\n",
+                "\n",
+                "        # if not zero subtract min(top10)\n",
+                "        test.loc[test['id'] == i, 'predictions'] = test.loc[test['id'] == i, 'predictions'].apply(lambda x: x - max(top10) + 10 if x != 0 else 0)\n",
+                "\n",
+                "        # if 10 set to 12\n",
+                "        test.loc[test['id'] == i, 'predictions'] = test.loc[test['id'] == i, 'predictions'].apply(lambda x: 12 if x == 10 else x)\n",
+                "        test.loc[test['id'] == i, 'predictions'] = test.loc[test['id'] == i, 'predictions'].apply(lambda x: 10 if x == 9 else x)\n",
+                "\n",
+                "    test['actual'] = test['rank'] + 1\n",
+                "\n",
+                "\n",
+                "    accuracy = accuracy_score(test['actual'], test['predictions'])\n",
+                "    print(\"Accuracy: %.2f%%\" % (accuracy * 100.0))\n",
+                "\n",
+                "    return test"
+            ]
+        },
+        {
+            "cell_type": "code",
+            "execution_count": null,
+            "metadata": {},
+            "outputs": [],
+            "source": [
+                "model_ranked, test_data, train_data = xgboost_rank_model(df_xgboost.loc[df_xgboost['points'] > 0], seed=7, test_size=0.33)\n",
+                "out = ranked_model_predictions(model_ranked, test_data)"
+            ]
+        },
+        {
+            "cell_type": "code",
+            "execution_count": null,
+            "metadata": {},
+            "outputs": [],
+            "source": [
+                "model_ranked_all, test_data_all, train_data_all = xgboost_rank_model(df_xgboost, seed=7, test_size=0.33)\n",
+                "out_all = ranked_model_predictions(model_ranked_all, test_data_all)"
+            ]
+        },
```

Comment:
this is a very good comparison. Interesting that accuracy drops that much when 0's are excluded (though I think we did talk about this at some point). Also, what I don't quite understand is why the accuracy in the basic XGBoost is still 20% higher than in the ranked including 0's.